### PR TITLE
Feature/create instance hierarchy

### PR DESCRIPTION
**Pull request by CarlHembrough**
### What

Created utilities to: 
* extract a hierarchy from a V4 file
* output the hierarchy as cypher script that creates a generic hierarcy
* clone the generic hierarchy and create an instance hierarchy

The main output of the task as this stage was a way to manually build an instance hierarchy. I have added details of how to do this to the readme.

I have added the graph DB design details to the existing design doc: https://docs.google.com/document/d/1XcU9B9KPVdh5OdHsqNd9fhcFgFHuGCdjchc90ZcChWI/edit#

### How to review

* Try the instructions to create the instance hierarchy.
* Review the updates to the design doc and check they make sense.
* Review the query
* maybe a light review of the code - as they are not to be used in prod

### Who can review

Anyone

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,6 +3,31 @@ dp-hierarchy-builder
...
+* Create the generic hierarchy - please note you will have to replace the value for 'instanceid' at the start
+
+`instanceid="c4-d39e-4e2f-942b-3ca105" && cat cmd/builder/build.cypher | sed "s/12345/$instanceid/g" | cypher-shell`
+
+#### Query the instance hierarchygit s
+
+* get children of a node - note that the "code" field will have to be set in the query
+
+``` 
+MATCH (n:\`_hierarchy_node_c4-d39e-4e2f-942b-3ca105_Aggregate\` {code:"cpi1dim1T120000"})<-[r:hasParent]-(child) RETURN child`
```
Comment (Matt-Guest) [inline README.md]: Do you need the escape character here? The other queries don't have them?

File: cmd/builder/main.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+var codeListID = flag.String("code-list-id", "e44de4c4-d39e-4e2f-942b-3ca10584d078", "")
+var instanceID = flag.String("instance-id", "12345", "")
+var dimensionName = flag.String("dimension-name", "Aggregate", "")
+
+func main() {
+
+	driver := bolt.NewDriver()
+
+	connection, err := driver.OpenNeo("bolt://localhost:7687")
```
Comment (Matt-Guest) [inline cmd/builder/main.go]: Is it worth making the neo4j url a command line option?

File: cmd/builder/main.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+	if err != nil {
+		return
+	}
+	defer connection.Close()
+
+	createInstanceHierarcyConstraints(connection)
+	if err != nil {
+		log.Println(err.Error())
+		return
```
Comment (Matt-Guest) [inline cmd/builder/main.go]: Is it worth doing `os.exit(1)` for all/most error events in the main function?

File: cmd/builder/main.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	createConstraint := fmt.Sprintf("CREATE CONSTRAINT ON (n:`_hierarchy_node_%s_%s`) ASSERT n.code IS UNIQUE;", *instanceID, *dimensionName)
+	stmtInsert, err := connection.PrepareNeo(createConstraint)
+	if err != nil {
+		return err
+	}
+	results, err := stmtInsert.ExecNeo(nil)
+	if err != nil {
+		return err
+	}
+	fmt.Println(results)
```
Comment (Matt-Guest) [inline cmd/builder/main.go]: Any reason why this is using `fmt` instead of `log`

File: cmd/builder/main.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+	if err != nil {
+		return err
+	}
+
+	results, err := stmtInsert.ExecNeo(nil)
+	if err != nil {
+		stmtInsert.Close()
+		return err
+	}
+	fmt.Println(results)
```
Comment (Matt-Guest) [inline cmd/builder/main.go]: See top comment

File: cmd/builder/main.go [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+		"MERGE (node)-[r:hasParent]->(parent);", *codeListID, *codeListID, *instanceID, *dimensionName, *instanceID, *dimensionName)
+	log.Println(insert)
+
+	stmtInsert, err := connection.PrepareNeo(insert)
+	if err != nil {
+		return err
+	}
+	results, err := stmtInsert.ExecNeo(nil)
+	if err != nil {
+		stmtInsert.Close()
```
Comment (Matt-Guest) [inline cmd/builder/main.go]: Could `ExecNeo` return a nil statement on err?

File: cmd/generator/main.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+func traverseNodesWriteCypher(nodes []*v4.HierarchicalDimensionOption, buffer *bytes.Buffer, parent *v4.HierarchicalDimensionOption) {
+	for _, node := range nodes {
+
+		if parent != nil {
+			buffer.WriteString(",\n")
+		}
+
+		//(cpi1dim1S50602:`_e44de4c4-d39e-4e2f-942b-3ca10584d078_hierarchy_node` { code:"cpi1dim1S50602" })
```
Comment (Matt-Guest) [inline cmd/generator/main.go]: Can this line be removed?

File: cmd/generator/v4/reader.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+		record, err = reader.csvReader.Read()
+
+		if err != nil {
+			return "", "", err
+		}
+
+		code = record[reader.codeIndex]
+		_, codeAlreadySeen = codesRead[code]
+		//fmt.Println(codesRead, codeAlreadySeen)
```
Comment (Matt-Guest) [inline cmd/generator/v4/reader.go]: commented out line

File: cmd/generator/v4/hierarchical_reader.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
+package v4
+
+import "strings"
+
+type HierarchicalLabelReader struct {
```
Comment (Matt-Guest) [inline cmd/generator/v4/hierarchical_reader.go]: does this structure need godocs?

File: README.md (hunk 1 above)
Comment (CarlHembrough) [reply README.md]: Nope - removed them

File: cmd/builder/main.go (hunk 2 above)
Comment (CarlHembrough) [reply cmd/builder/main.go]: Done

File: cmd/builder/main.go (hunk 3 above)
Comment (CarlHembrough) [reply cmd/builder/main.go]: Done

File: cmd/builder/main.go (hunk 4 above)
Comment (CarlHembrough) [reply cmd/builder/main.go]: Changed to use log

File: cmd/builder/main.go (hunk 5 above)
Comment (CarlHembrough) [reply cmd/builder/main.go]: done

File: cmd/builder/main.go (hunk 6 above)
Comment (CarlHembrough) [reply cmd/builder/main.go]: The statement is not returned from ExecNeo - its created before from `connection.PrepareNeo(insert)` so should be OK to call close if the exec fails.

File: cmd/generator/main.go (hunk 7 above)
Comment (CarlHembrough) [reply cmd/generator/main.go]: done

File: cmd/generator/v4/hierarchical_reader.go (hunk 9 above)
Comment (CarlHembrough) [reply cmd/generator/v4/hierarchical_reader.go]: Added godocs

File: cmd/generator/v4/reader.go (hunk 8 above)
Comment (CarlHembrough) [reply cmd/generator/v4/reader.go]: Removed.

**Pull request merged**